src/cancel_order.rs

```diff
@@ -1,6 +1,198 @@
+use crate::big_decimal::{BigDecimal, WRatio};
+use crate::ref_finance::ref_finance;
+use crate::ref_finance::{Action, SwapAction, TokenReceiverMessage};
+use crate::utils::NO_DEPOSIT;
+use crate::utils::{ext_market, ext_token};
 use crate::*;
+use near_sdk::env::{current_account_id, signer_account_id};
+use near_sdk::{ext_contract, is_promise_success, log, Gas, PromiseResult};
+
+#[ext_contract(ext_self)]
+trait ContractCallbackInterface {
+    fn remove_liquidity_callback(&self, order_id: U128, swap_fee: U128, price_impact: U128);
+    fn cancel_swap_callback(&mut self, order_id: U128, swap_fee: U128, price_impact: U128);
+    fn market_data_callback(
+        &mut self,
+        order_id: U128,
+        swap_fee: U128,
+        price_impact: U128,
+        market_data: Option<MarketData>,
+    );
+}
 
 #[near_bindgen]
 impl Contract {
-    fn cancel_order(&mut self, order_id: U128, swap_fee: U128, price_impact: U128) {}
+    fn cancel_order(&mut self, order_id: U128, swap_fee: U128, price_impact: U128) {
+        let orders = self.orders.get(&signer_account_id()).unwrap_or_else(|| {
+            panic!("Orders for account: {} not found", signer_account_id());
+        });
+
+        let order = orders.get(&(order_id.0 as u64)).unwrap_or_else(|| {
+            panic!("Order with id: {} not found", order_id.0);
+        });
+
+        //TODO: set real min_amount_x/min_amount_y
+        let amount = 1;
+        let min_amount_x = order.amount;
+        let min_amount_y = 0;
+
+        if order.status == OrderStatus::Pending {
+            ref_finance::ext(self.ref_finance_account.clone())
+                .with_static_gas(Gas(10))
+                .with_attached_deposit(1)
+                .remove_liquidity(
+                    order.lpt_id.clone(),
+                    U128(amount),
+                    U128(min_amount_x),
+                    U128(min_amount_y),
+                )
+                .then(
+                    ext_self::ext(current_account_id())
+                        .with_static_gas(Gas(5))
+                        .with_attached_deposit(NO_DEPOSIT)
+                        .remove_liquidity_callback(order_id, swap_fee, price_impact),
+                );
+        } else {
+            self.swap(order_id, swap_fee, price_impact);
+        }
+    }
+
+    #[private]
+    pub fn remove_liquidity_callback(&self, order_id: U128, swap_fee: U128, price_impact: U128) {
+        require!(is_promise_success(), "Some problem with remove liquidity");
+        self.swap(order_id, swap_fee, price_impact);
```

Comment:
@TymofiiSavinov we actually should jump right to `cancel_swap_callback` here as we already have all USDT

Reply:
we have swap_callback in create order, but with not same functionality. I changed name to order_cancel_swap_callback